File: backend/src/utils/initDb.ts
import { Client } from 'pg';
import dotenv from 'dotenv';
import fs from 'fs';
import path from 'path';

dotenv.config({ path: path.join(__dirname, '../../../.env') });

const setup = async () => {
    // Prefer DATABASE_URL if present
    const connectionString = process.env.DATABASE_URL;

    if (!connectionString) {
        console.error('DATABASE_URL not found in .env (expected for local Postgres).');
        process.exit(1);
    }

    console.log('Connecting to database to apply schema...');

    const isLocal = /localhost|127\.0\.0\.1/i.test(connectionString);

    const client = new Client(
        isLocal
            ? { connectionString }
            : {
                  connectionString,
                  ssl: {
                      rejectUnauthorized: false,
                  },
              }
    );

    try {
        await client.connect();

        console.log('Applying schema...');
        const schemaPath = path.join(__dirname, '../../schema.sql');
        const schemaSql = fs.readFileSync(schemaPath, 'utf8');

        await client.query(schemaSql);
        console.log('Schema applied successfully.');
        await client.end();
    } catch (err) {
        console.error('Database setup failed:', err);
        process.exit(1);
    }
};

setup();

File: backend/init_db.py
"""Run schema.sql against the database configured in .env"""
import asyncio
import asyncpg
import os
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))

async def run():
    dsn = os.environ.get('DATABASE_URL')
    if not dsn:
        print("ERROR: DATABASE_URL not set in .env")
        return
    # asyncpg doesn't support channel_binding as a DSN parameter
    dsn = dsn.split('&channel_binding')[0]
    print(f"Connecting to database...")
    conn = await asyncpg.connect(dsn, ssl='require')
    
    schema_path = os.path.join(os.path.dirname(__file__), 'schema.sql')
    sql = open(schema_path, 'r').read()
    await conn.execute(sql)
    print("Schema applied successfully!")
    await conn.close()

if __name__ == '__main__':
    asyncio.run(run())
